src/app/codegen-interaction-model/Model.cpp

```diff
@@ -0,0 +1,267 @@
+                                                       uint16_t start_index)
+{
+    for (unsigned i = start_index; i < endpoint->clusterCount; i++)
+    {
+        const EmberAfCluster & cluster = endpoint->cluster[i];
+        if (!IsServerMask(cluster.mask & CLUSTER_MASK_SERVER))
+        {
+            continue;
+        }
+
+        return ClusterEntryFrom(endpointId, cluster);
+    }
+
+    return InteractionModel::ClusterEntry::Invalid();
+}
+
+/// Load the cluster information into the specified destination
+void LoadAttributeInfo(const ConcreteAttributePath & path, const EmberAfAttributeMetadata & attribute,
+                       InteractionModel::AttributeInfo * info)
+{
+    if (attribute.attributeType == ZCL_ARRAY_ATTRIBUTE_TYPE)
+    {
+        info->flags.Set(InteractionModel::AttributeQualityFlags::kListAttribute);
+    }
+
+    // TODO: Set additional flags:
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kChangesOmitted)
+}
+
+InteractionModel::AttributeEntry AttributeEntryFrom(const ConcreteClusterPath & clusterPath,
+                                                    const EmberAfAttributeMetadata & attribute)
+{
+    InteractionModel::AttributeEntry entry;
+
+    entry.path = ConcreteAttributePath(clusterPath.mEndpointId, clusterPath.mClusterId, attribute.attributeId);
+    LoadAttributeInfo(entry.path, attribute, &entry.info);
+
+    return entry;
+}
+
+} // namespace
+
+CHIP_ERROR Model::ReadAttribute(const InteractionModel::ReadAttributeRequest & request, InteractionModel::ReadState & state,
+                                AttributeValueEncoder & encoder)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR Model::WriteAttribute(const InteractionModel::WriteAttributeRequest & request, AttributeValueDecoder & decoder)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR Model::Invoke(const InteractionModel::InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
+                         InteractionModel::InvokeReply & reply)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+EndpointId Model::FirstEndpoint()
+{
+    // find the first enabled index
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+    for (uint16_t endpointIndex = 0; endpointIndex < lastEndpointIndex; endpointIndex++)
+    {
+        if (emberAfEndpointIndexIsEnabled(endpointIndex))
+        {
+            return emberAfEndpointFromIndex(endpointIndex);
+        }
+    }
+
+    // No enabled endpoint found. Give up
+    return kInvalidEndpointId;
+}
+
+EndpointId Model::NextEndpoint(EndpointId before)
+{
+    // find the first enabled index
+    bool beforeFound = false;
+
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+    for (uint16_t endpointIndex = 0; endpointIndex < lastEndpointIndex; endpointIndex++)
```

Comment:
This will be an O(N^2) linear search cost unless we mandate "must be in-order". Same for other elements. Based on my profiling of wildcard expansion, this will be very costly. The iteration must be able to allow O(N) cost iteration of the N entries in the common case, rather than needing possibly several O(N^2)